### ISSUE TITLE
Intro on how to add multiple language support?

### ISSUE THREAD

**Issue by dooniem** [tdalon/ahk#16]
Hi, as many other it seems, I came here for the mute-button. Appreciate the great work you have laid down on this one!
So I copied "Teams_Mute" and assigned the function to a shortcut instead. I copied the necessary extensions, converted it into an executable I now run on Windows startup. This works fine for "English Teams".

So now my remaining task is to make the code work for 2 languages. But my coding skills isn't as good as yours.
I'm not asking you to rewrite all the code on my behalf, but it would be helpful if you gave me some pointers?

My first error is "Could not find Meeting Window! , No active Teams meeting window found!", which I'm guessing comes from the function `Teams_GetMeetingWindow.` The only language sensitive code there runs just if Maximize:=true? 
And the specific text it is looking for is _"Navigate back to call window."_. I can't seem to find where I should look for the translated version of this text. Can't find that button.

But next I guess I should change 
`El :=  TeamsEl.FindFirstByNameAndType("Mute", "button",,2)`
and 
`El :=  TeamsEl.FindFirstByNameAndType("Unmute", "button",,2)`

I tried changing "Mute" to "Demp" which is the Norwegian equivalent, but that wasn't enough it seems. So I guess more changes are needed.

But since I want to support 2 languages in the same code, I was hoping you had some pointers for me on how to code this?
The way of least resistance would be to make a new "FindFirstByNameAndType"-search if the first one returns empty. But I don't know how to structure the variable or code for this.
I guess the other way to do it is to check for the locale code and replace "Mute" with varMute and assign the correct string for varMute based on the discovered locale code. But then you probably have to implement a getLang function...That's outside my knowledge.

So long story short, could you help me on my way making the functions support at least 2 languages?

**Comment by tdalon** [comment]
Hi dooniem. First of all, thanks for your interest in my code.
I have tried to answer your question (sorry for the quick shot) here:
https://tdalon.blogspot.com/2023/08/teams-powertools-lang-extension.html

I hope it is clear enough.
If you want me to implement your language version in my repo/ Library just ping me back with your updated code.

Instead of extracting what you want from my code, why not use it as is. If you don't use all the features it is fine.
With the PowerTool for example you could use the options to assign an hotkey or also use the Launcher (though you can also install the Teamsy Launcher only.)

Kind Regards,
Thierry

**Comment by dooniem** [comment]
Thanks for taking the time to reply this fast.
I decided to try and answer your last question before reading your blog post about translation.
The main reason is actually because I don't fully understand the contents of your PowerTools, if it's a "suite" of overkill or if it's something I'd like to start using. This is mainly because most of your videos, for example on "Teamsy", is marked as "This video is unavailable". Since your "sales pitch" is mostly through videos and not text and images, I can't get a grip on what all your coding does and if it benefits me or my team. There's a lot of add-ons, lot of new things for me and my team to try and understand. 

What I could understand though, was that this PowerTool makes it easier to do "Meeting reactions". Which honestly isn't something I'm using nor see the need to implement as of now.

I also understood, through reading comments elsewhere and trying to grasp your code, that you had solved my main task. To be able to mute/un-mute Teams whether or not Teams was my active window.

So I concluded that instead of installing a new software with an installer (elevated permissions needed?), and a to-do-list after installing. The ease of having a compiled AHK.exe with the shortcut pre-loaded, put into the Windows startup folder, was easier to share and distribute to my colleagues (no admin privileges)  than downloading "a whole suite of software/scripts" they had to navigate through before being able to mute their mic.

**Comment by dooniem** [comment]
I've tried implementing what's on your blogpost, thanks for going with Teams_GetLang as suggested 😅 
That function works. So the code is able to set the correct MuteName and Unmutename. But as I mentioned, the code doesn't work simply by replacing the English names with the language specific ones.

Even though a meeting is active, `WinId := Teams_GetMeetingWindow() ` returns empty when I've changed the display language of Teams. So there probably needs to be some Tweaking with the `WinID` as well?

**Comment by tdalon** [comment]
I have implemented and tested the mute function for the German language setting. See updated [blog post](https://tdalon.blogspot.com/2023/08/teams-powertools-lang-extension.html) with a short YouTube video.
I had somehow an issue / it wasn't working immediately as expected. (I think after a Teams Client restart because of lang setting change one also need to restart the AHK script.)
But at the end I got it working.
I think, you forgot to update the Teams_GetMeetingWindow and Teams_IsMeetingWindow functions.
Updated code was also pushed in the repository.
Give it a try now with the new implementation and tell me if you could get it work for your language. (If you tell me which second lang you have I can also implement this for you but the purpose of teaching you how to do it would be lost.)

**Comment by tdalon** [comment]
> Thanks for taking the time to reply this fast. I decided to try and answer your last question before reading your blog post about translation. The main reason is actually because I don't fully understand the contents of your PowerTools, if it's a "suite" of overkill or if it's something I'd like to start using. This is mainly because most of your videos, for example on "Teamsy", is marked as "This video is unavailable". Since your "sales pitch" is mostly through videos and not text and images, I can't get a grip on what all your coding does and if it benefits me or my team. There's a lot of add-ons, lot of new things for me and my team to try and understand.
> 
> What I could understand though, was that this PowerTool makes it easier to do "Meeting reactions". Which honestly isn't something I'm using nor see the need to implement as of now.
> 
> I also understood, through reading comments elsewhere and trying to grasp your code, that you had solved my main task. To be able to mute/un-mute Teams whether or not Teams was my active window.
> 
> So I concluded that instead of installing a new software with an installer (elevated permissions needed?), and a to-do-list after installing. The ease of having a compiled AHK.exe with the shortcut pre-loaded, put into the Windows startup folder, was easier to share and distribute to my colleagues (no admin privileges) than downloading "a whole suite of software/scripts" they had to navigate through before being able to mute their mic.

Yes. I know,  a pitch/demo video about the tools is missing. If you have interest I could make a demo for you and record it.

Basically this isn't a big complicated software to install etc. 
These are simply small [AHK-based tools](https://tdalon.github.io/ahk/PowerTools-Setup) that you can deploy/use also as [compiled exe](https://tdalon.github.io/ahk/PowerTools-Setup#standalone-executables).
My proposal was that you could just take for example the TeamsShortcuts.ahk including Libraries or the standalone TeamsShortcuts. .exe out of the box (instead of extracting functions from the library). You would have also the launcher included (more convenient than hotkeys) or you could also configure the [hotkeys](https://tdalon.github.io/ahk/Teams-Global-Hotkeys) etc.

I understand if you plan only to use the mute function it might be better to extract this part yourself.

**Comment by dooniem** [comment]
Ah I see where I failed, I searched the document for `FindFirstByNameAndType` and not `FindFirstByName` hence didn't find the changes you made in the video...Now it works! Thanks for basically spoon feeding me the right way to go! 🥇 

I see you made a discrepancy in the code. While the English code search for "Calling controls", the German code search for the equivalent of "Meeting controls". Two different controls. No biggie, probably works a lifetime, but in case you care for such things...  

For anyone who come across this later, I decided to call `Lang := Teams_GetLang()` only once at the start of muting, then set `global ` for the variables outside of the main function.
In the blogpost all language edits needed but this one is mentioned: `global NavigateBack :="Navigate back to call window."`. So I just thought I'd throw it out here.

**Comment by tdalon** [comment]
Thanks for the hint with the wrong Calling controls. I have fixed this.
Thanks also for the hint with the global variable. I have improved the GetLang function to memorize the lang and not read the file each time. It shall make your use of the global variable unnecessary.
I will close the issue then.

**Comment by dooniem** [comment]
No problem and thanks for developing this once again. I guess I'll check in after a while to see where this project is getting to.

Just thought I'd mention, in case you haven't thought about it. When making getLang static, the code will fail when the script is loaded and the user change his/her language. The script needs to be reloaded  for each language switch. Maybe your launcher actually do reload the code on each language switch 😅 

The smoothest approach would be to catch specific errors, if an error is caught, then ignore the static Lang value and re-read the actual language

**Comment by tdalon** [comment]
> Just thought I'd mention, in case you haven't thought about it. When making getLang static, the code will fail when the script is loaded and the user change his/her language. The script needs to be reloaded for each language switch. Maybe your launcher actually do reload the code on each language switch 😅
> 
Yes. if you change the Teams client lang setting, you will have to restart the Teams AHK scripts.
This is ok for me. Normally you don't change the lang setting more than once - except for debugging these scripts ;-)

**Comment by tdalon** [comment]
@dooniem I have updated the language handling see recent post https://tdalon.blogspot.com/2023/10/teams-powertools-lang.html
Also today I have refactored a bit the getlangname function
